### PR TITLE
POC: Refactored storage to use flysystem

### PR DESCRIPTION
**Pull request by wachterjohannes**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | no
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | none
| Related issues/PRs | none
| License | MIT
| Documentation PR | none

#### What's in this PR?

Refactor Storage to use flysystem filesystem abstraction.

#### Why?

This will make the cloud deployment easier.

### CODE REVIEW

**Comment by wachterjohannes** [comment]
@alexander-schranz  what do you think? not tested excessively but i think something like this should work

**Comment by alexander-schranz** [comment]
@wachterjohannes the interface for filesystem is defined in https://github.com/sulu/sulu/pull/4024

File: src/Sulu/Bundle/MediaBundle/Media/Storage/StorageInterface.php [hunk 1]
```diff
@@ -19,45 +19,23 @@
...
      * Save the document in a storage and give back the path to the document.
-     *
-     * @param $tempPath
-     * @param $fileName
-     * @param $version
-     *
-     * @return mixed
      */
-    public function save($tempPath, $fileName, $version);
+    public function save(string $tempPath, string $fileName, int $version): string;
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/Media/Storage/StorageInterface.php]: storage should not know about versions thats why we removed it in #4024 https://github.com/sulu/sulu/pull/4024/files#diff-d1603c50279a10b85948f4e515ee33c4

File: src/Sulu/Bundle/MediaBundle/Media/Storage/StorageInterface.php [hunk 2]
```diff
@@ -19,45 +19,23 @@
...
-     * @param $storageOption
-     *
-     * @return string
+     * Returns the content for the given file as a resource.
      *
-     * @deprecated Deprecated since 1.4, will be removed in 2.0
+     * @return resource
      */
-    public function load($fileName, $version, $storageOption);
+    public function loadAsStream(string $fileName, int $version, string $storageOption);
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/Media/Storage/StorageInterface.php]: only storageOptions should be needed to load the image see https://github.com/sulu/sulu/pull/4024/files#diff-d1603c50279a10b85948f4e515ee33c4

File: src/Sulu/Bundle/MediaBundle/Media/Storage/LocalStorage.php [hunk 3]
```diff
@@ -11,12 +11,13 @@
 
 namespace Sulu\Bundle\MediaBundle\Media\Storage;
 
-use stdClass;
+use League\Flysystem\FileExistsException;
+use League\Flysystem\FileNotFoundException;
+use League\Flysystem\Filesystem;
```
Comment (alexander-schranz) [inline src/Sulu/Bundle/MediaBundle/Media/Storage/LocalStorage.php]: would not use Flysystem in the LocalStorage there should be a FlySystemStorage which implements `ExternalStorageInterface` to get the external urls.

File: composer.json [hunk 4]
```diff
@@ -32,18 +32,19 @@
...
         "guzzlehttp/guzzle": "^6.2",
         "guzzlehttp/psr7": "^1.4",
-        "php-http/client-common": "^1.1.0",
-        "php-http/guzzle6-adapter": "^1.1",
         "imagine/imagine": "~0.6.1 || ~0.7.0",
-        "jms/serializer-bundle": "^2.2",
         "jms/serializer": "^1.13",
+        "jms/serializer-bundle": "^2.2",
         "layershifter/tld-extract": "^1.1",
+        "league/flysystem": "^1.0",
```
Comment (alexander-schranz) [inline composer.json]: for me flysystem should be optional dependency.

**Comment by wachterjohannes** [comment]
replaced by https://github.com/sulu/sulu/pull/4264 and https://github.com/sulu/sulu/pull/4265

**Pull request closed without merging**